atr/docs/user-interface.md

```diff
@@ -46,30 +46,104 @@ Template rendering happens in a thread pool to avoid blocking the async event lo
+- `form.Int` - number input
+- `form.StrList` - multiple checkboxes that collect strings
+- `form.File` - single file upload
+- `form.FileList` - multiple file upload
+- `form.Enum[EnumType]` - dropdown select from enum values
+- `form.Set[EnumType]` - multiple checkboxes from enum values
 
-To use a form in a route, create it with `await FormClass.create_form()`. For POST requests, pass `data=await quart.request.form` to populate it with the submitted data. Then validate with `await form.validate_on_submit()`. If validation passes, you extract data from `form.field_name.data` and proceed. If validation fails, re-render the template with the form object, which will then display error messages.
+Fields are required by default. To make a field optional, use Python's standard `Optional` or union syntax: `field_name: str | None = None`.
 
-The [`forms`](/ref/atr/forms.py) module also provides rendering functions that generate Bootstrap-styled HTML. The function [`forms.render_columns`](/ref/atr/forms.py:render_columns) creates a two-column layout with labels on the left and inputs on the right. The function [`forms.render_simple`](/ref/atr/forms.py:render_simple) creates a simpler vertical layout. The function [`forms.render_table`](/ref/atr/forms.py:render_table) puts the form inside a table. All three functions return htpy elements, which you can embed in templates or return directly from route handlers.
+The `widget` parameter in [`form.label`](/ref/atr/form.py:label) lets you override the default widget for a field type. Available widgets include: `TEXTAREA`, `CHECKBOXES`, `SELECT`, `RADIO`, `HIDDEN`, and others from the `form.Widget` enum.
+
+### Using Forms in Routes
+
+To use a form in a route, validate the form data from a POST request using [`form.validate`](/ref/atr/form.py:validate):
+```python
+@post.committer("/keys/add")
+@post.form(shared.keys.AddOpenPGPKeyForm)
+async def add(session: web.Committer, add_openpgp_key_form: shared.keys.AddOpenPGPKeyForm) -> web.WerkzeugResponse:
+    """Add a new public signing key to the user's account."""
+    try:
+        key_text = add_openpgp_key_form.public_key
+        selected_committee_names = add_openpgp_key_form.selected_committees
+
+        # Process the validated form data...
+        async with storage.write() as write:
+            # ...
+
+        await quart.flash("OpenPGP key added successfully.", "success")
+    except web.FlashError as e:
+        await quart.flash(str(e), "error")
+    except Exception as e:
+        log.exception("Error adding OpenPGP key:")
+        await quart.flash(f"An unexpected error occurred: {e!s}", "error")
+
+    return await session.redirect(get.keys.keys)
+```
+
+The [`form.validate`](/ref/atr/form.py:validate) function takes the form class, the form data dictionary, and an optional context dictionary. If validation succeeds, it returns an instance of your form class with validated data. If validation fails, it raises a `pydantic.ValidationError`.
```

Comment:
This is true, but as noted above, `form.validate` is not used in the sample code and should be avoided except in certain circumstances.